### PR TITLE
Implement birdhouse bird spawning [OLB-51]

### PR DESCRIPTION
**Pull request by jamestjw**
- Birdhouse spawns birds every 50 ticks
- Implement ticker to manage server ticks
- Refactor seed consumption code to use ticker

### CODE REVIEW

**Comment by jamestjw** [comment]
The x,y,z offset thing still seems to be a bit broken, but this is simply because the MC API is not documented at all, I don't even know what the method parameters are soooooo yeah

EDIT: Resolved

File: build.gradle [hunk 1]
```diff
@@ -18,6 +20,7 @@ apply plugin: 'eclipse'
 apply plugin: 'maven-publish'
 apply plugin: 'java'
 apply plugin: 'checkstyle'
+apply plugin: 'com.adarshr.test-logger'
```
Comment (jamestjw) [inline build.gradle:24]: This plugin makes it easier to look at the results of unit tests
image

File: config/checkstyle/checkstyle.xml [hunk 2]
```diff
@@ -391,12 +391,6 @@
     </module>
 
     <module name="InvalidJavadocPosition"/>
-
-    <module name="MissingJavadocMethodCheck">
-      <property name="minLineCount" value="1"/>
-      <property name="allowMissingPropertyJavadoc" value="true"/>
-      <property name="ignoreMethodNamesRegex" value="(set.*|get.*)"/>
-    </module>
```
Comment (jamestjw) [inline config/checkstyle/checkstyle.xml:399]: I found this to be silly, I don't believe in the "every method needs to be documented" philosophy

**Review by shafiqimtiaz — DISMISSED** [review]
Looks good 👍🏽

File: src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java [hunk 3]
```diff
@@ -25,7 +32,8 @@
     private final ItemStackHandler itemHandler = createHandler();
     private final LazyOptional<IItemHandler> handler = LazyOptional.of(() -> itemHandler);
 
-    private int counter;
+    private Ticker tickerForBirdSpawns = new FixedIntervalTicker(150);
+    private Ticker tickerForSeedConsumption = new FixedIntervalTicker(100);
```
Comment (jamestjw) [inline src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java:37]: Handling the ticking using a helper class seemed to be a good idea (separation of concerns)

File: src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java [hunk 4]
```diff
@@ -69,21 +75,49 @@ public void load(CompoundTag tag) {
...
+        // Move the spawned bird to the destination
+        // 0.0 because it works, not sure what this supposed to be
+        newBird.moveTo(desBlockPos, (float) 0.0, (float) 0.0);
+    }
+
+    private void handleBirdSpawnForTick() {
+        if (tickerForBirdSpawns.tick()) {
+            int xOffset = ThreadLocalRandom.current().nextInt(-20, 20);
+            int yOffset = 50;
+            int zOffset = ThreadLocalRandom.current().nextInt(-20, 20);
```
Comment (jamestjw) [inline src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java]: The y offset doesn't seem to work, I think its possible that birds have to be spawned on the ground (can you guys test this out?)

File: src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java [hunk 5]
```diff
@@ -69,21 +75,49 @@ public void load(CompoundTag tag) {
...
         }
 
         // When adding seeds, if the seeds are "burnable" then set the block to active.
         BlockState blockState = getBlockState();
-        if (blockState.getValue(BlockStateProperties.CONDITIONAL) != counter > 0) {
-            level.setBlock(worldPosition, blockState.setValue(BlockStateProperties.CONDITIONAL, counter > 0),
+        if (blockState.getValue(BlockStateProperties.CONDITIONAL) != seedExtracted) {
+            level.setBlock(worldPosition, blockState.setValue(BlockStateProperties.CONDITIONAL, seedExtracted),
                     Block.UPDATE_ALL);
         }
```
Comment (jamestjw) [inline src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java]: @Zachary-Concordia can you verify that I havent altered the behaviour of the birdhouse as far as seed consumption is concerned?

**Comment by jamestjw** [comment]
> Looks good 👍🏽

@shafiqimtiaz  Did you try running it? You're welcome to tweak the offsets at which birds get spawned (there might be a more ideal one)

**Comment by shafiqimtiaz** [comment]
@jamestjw 
- The bird spawning works (but dont know where is it spawning from, will try tweaking the offsets)
Note: I think it would be great if the bird spawner could be kept as a separate block.
- I think the birdhouse should spawn seeds, which the player can pick up and feed birds. Also, the birdhouse should only have the feature of attracting birds. I'll try to work on it.

- Issue found: The birdhouse now doesn't consume seeds

**Comment by jamestjw** [comment]
> @jamestjw
> 
> * The bird spawning works (but dont know where is it spawning from, will try tweaking the offsets)
>   Note: I think it would be great if the bird spawner could be kept as a separate block.
> * I think the birdhouse should spawn seeds, which the player can pick up and feed birds. Also, the birdhouse should only have the feature of attracting birds. I'll try to work on it.
> * Issue found: The birdhouse now doesn't consume seeds

@shafiqimtiaz As for where the birds are being spawned, they should be spawned randomly around the birdhouse within a 20 unit radius. Could you explain the merits of putting the spawner in a separate block?

The spawning of seeds is an idea I guess, but bring it up with the others elsewhere, since it is unrelated to this pull request.

I will investigate the issue regarding seed consumption.

**Comment by shafiqimtiaz** [comment]
@jamestjw the merits of keeping the spawner separate of birdhaus is primarily in order to test the functionality of birdhaus. The birdhaus attracts birds, the spawner will help test the fact whether the birdhaus does so. But I guess, we can discuss this issue in the meeting. :)

**Comment by jamestjw** [comment]
> @jamestjw the merits of keeping the spawner separate of birdhaus is primarily in order to test the functionality of birdhaus. The birdhaus attracts birds, the spawner will help test the fact whether the birdhaus does so. But I guess, we can discuss this issue in the meeting. :)

@shafiqimtiaz I see, but as far as this task is concerned, spawning from the birdhouse is what we talked about during our previous meeting. So yeah we could further discuss in the future to make it work differently. Though I must say that I feel like spawning birds from the birdhouse itself isn't too bad of an idea, since it gives the user the impression that birds are indeed somehow 'attracted' to the presence of a birdhouse.

File: src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java (hunk 5 above)
Comment (jamestjw) [reply src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java]: Ok I think I screwed it up, okay so correct me if I'm wrong, here's my new interpretation of what your code was doing

1. Every 100 ticks, the block becomes ready to consume a seed (here's when a block becomes active)
2. Once it does consume a seed, it becomes inactive until another 100 ticks pass

Comment (Zachary-Concordia) [reply src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java]: More or less, yes. When the birdhouse consumes a seed, the counter is set to 100 and counts down one every tick. If it is 0 or less, it eats another item by calling the extract item from slot 0.  Every tick it also checks to see if the block state conditional bool is the same as the (counter > 0). If not, it updates the block state conditional to the new state of (counter > 0).

Comment (jamestjw) [reply src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java]: @Zachary-Concordia alright cool! I just changed the code to match the intended behaviour. I believe that code describes the intended behaviour more clearly now.

@shafiqimtiaz If you have a moment, could you retest this :) I think we got back the original behaviour

File: src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java (hunk 4 above)
Comment (Zachary-Concordia) [reply src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java]: Will test this out tonight

File: src/main/java/com/ocelotslovebirds/birdhaus/ticker/Ticker.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package com.ocelotslovebirds.birdhaus.ticker;
+
+// TODO: Should the class have a different name?
```
Comment (Zachary-Concordia) [inline src/main/java/com/ocelotslovebirds/birdhaus/ticker/Ticker.java]: I can't think of a better name.

File: src/main/java/com/ocelotslovebirds/birdhaus/ticker/FixedIntervalTicker.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+    /**
+     * @param interval How often the ticker should tick
+     */
+    public FixedIntervalTicker(int interval) {
+        this.interval = interval;
+        this.counter = 0;
+    }
+
+    public boolean tick() {
```
Comment (Zachary-Concordia) [inline src/main/java/com/ocelotslovebirds/birdhaus/ticker/FixedIntervalTicker.java:15]: We could potentially shorten this to `return ++counter % interval == 0;`

File: src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java [hunk 8]
```diff
@@ -69,22 +78,61 @@ public void load(CompoundTag tag) {
...
-        if (counter <= 0) {
+    }
+
+    private void handleSeedsForTick() {
+        if (isActive) {
+            // If the birdhouse is active, try consuming a seed
             if (!itemHandler.extractItem(0, 1, false).isEmpty()) {
-                counter = 100;
-                setChanged();
```
Comment (Zachary-Concordia) [inline src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java:80]: Confirming that we don't need setchanged() for our current functionality. I forgot to remove it from the initial trying to get it to work. Good catch!

**Review by Zachary-Concordia — DISMISSED** [review]
Looks good, no major changes requested. Minor suggestion for the ticker class tick() function. Thanks!

File: src/main/java/com/ocelotslovebirds/birdhaus/ticker/FixedIntervalTicker.java (hunk 7 above)
Comment (jamestjw) [reply src/main/java/com/ocelotslovebirds/birdhaus/ticker/FixedIntervalTicker.java:15]: @Zachary-Concordia is it safe to infinitely increment a variable? I've never thought about this before

File: src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java (hunk 4 above)
Comment (chris7neves) [reply src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java]: You can set the y offset to 
this.getBlockPos().getY()

This will set the Y of the offset to the same Y level the birdhouse is on.

Comment (jamestjw) [reply src/main/java/com/ocelotslovebirds/birdhaus/blocks/BirdhouseBlockEntity.java]: @chris7neves hmm good point, I should try `this.getBlockPos().getY() + 20` or something to give the impression that the birds fly down from the sky

**Comment by jamestjw** [comment]
Ok perfect I got the y-offset to work, the birds spawn 10units above the birdhouses now

**Comment by chris7neves** [comment]
I think its good to go for a merge then. Because im realizing this PR modifies how ticking functions within the birdhouse block and the bird flocking latches onto that ticking, so i may need to modify how i go about changing the BirdhouseBlockEntity

**Comment by jamestjw** [comment]
> I think its good to go for a merge then. Because im realizing this PR modifies how ticking functions within the birdhouse block and the bird flocking latches onto that ticking, so i may need to modify how i go about changing the BirdhouseBlockEntity

@chris7neves yep that was precisely my concern when I wrote my comment in #9

**Review by DevV2M — APPROVED** [review]
All ok!

**Review by drouba — APPROVED** [review]

**Review by Zachary-Concordia — APPROVED** [review]
Approved

**Review by chris7neves — APPROVED** [review]
